stan/math/rev/fun/matrix_power.hpp

```diff
@@ -130,38 +25,49 @@ class matrix_product_vari : public vari {
  * @throw std::invalid_argument if the exponent is negative or the matrix is not
  * square.
  */
-template <typename EigMat, require_eigen_vt<is_var, EigMat>* = nullptr>
-inline Eigen::Matrix<value_type_t<EigMat>, EigMat::RowsAtCompileTime,
-                     EigMat::ColsAtCompileTime>
-matrix_power(const EigMat& M, const int n) {
-  using T = value_type_t<EigMat>;
-  constexpr int R = EigMat::RowsAtCompileTime;
-  constexpr int C = EigMat::ColsAtCompileTime;
-
-  check_square("matrix_power", "M", M);
-  if (n < 0)
-    invalid_argument("matrix_power", "n", n, "is ", ", but must be >= 0!");
-  if (M.rows() == 0)
-    invalid_argument("matrix_power", "M.rows()", M.rows(), "is ",
-                     ", but must be > 0!");
+template <typename T, require_rev_matrix_t<T>* = nullptr>
+inline plain_type_t<T> matrix_power(const T& M, const int n) {
   const auto& M_ref = to_ref(M);
+  check_square("matrix_power", "M", M);
```

Comment:
```suggestion
  check_square("matrix_power", "M", M_ref);
```

Reply:
I changed this around in a different way to that we check finite last (and only do a `to_ref` once we know we have a square, non-zero sized matrix).